setup_monitoring.sh

```diff
@@ -0,0 +1,107 @@
+  echo "📦 installing the Prometheus operator..."
+  export KUBECONFIG=$KUBECONFIG_HOST
+
+  oc process -f config/monitoring/prometheus.tmpl.yaml \
+    -p NAMESPACE="$TOOLCHAIN_OPERATOR_NS_HOST" \
+    -p SA_NAME=prometheus-k8s \
+    | oc apply -f -
+  
+  # needed by the `kube-rbac-proxy` sidecar to create TokenReviews and SubjectAccessReviews
+  oc adm policy add-cluster-role-to-user kube-rbac-proxy -z prometheus-k8s -n "$TOOLCHAIN_OPERATOR_NS_HOST"
+  
+  echo "✅ done installing Prometheus operator"
+  echo ""
+}
+
+######################################################################################
+## Deploying Grafana
+######################################################################################
+function deploy_grafana() {
+  echo "📦 deploying Grafana..."
+  # fetch the route to Thanos Querier on member cluster and 
+  export KUBECONFIG=$KUBECONFIG_MEMBER
+  THANOS_QUERIER_URL_MEMBER="https://$(oc get route/thanos-querier -n openshift-monitoring -o json | jq -r '.status.ingress[0].host')"
+  export THANOS_QUERIER_URL_MEMBER
+  echo "route to Thanos Querier on member cluster: $THANOS_QUERIER_URL_MEMBER"
+  
+  # create a `grafana` serviceaccount which is allowed to access Prometheus (via kube-rbac-proxy)
+  oc process -f config/monitoring/grafana_serviceaccount.tmpl.yaml \
+    -p NAMESPACE="$TOOLCHAIN_OPERATOR_NS_MEMBER" \
+    -p NAME=grafana \
+    | oc apply -f -
+  # needed to connect to Thanos Querier
+  oc adm policy add-cluster-role-to-user cluster-monitoring-view -z grafana -n "$TOOLCHAIN_OPERATOR_NS_MEMBER"
+  BEARER_TOKEN_MEMBER="$(oc serviceaccounts get-token grafana -n $TOOLCHAIN_OPERATOR_NS_MEMBER)"
+  export BEARER_TOKEN_MEMBER
+
+  # fetch the "local" route to Grafana on host cluster and retrieve SA token to use to connect to it
+  export KUBECONFIG=$KUBECONFIG_HOST
+  THANOS_QUERIER_URL_HOST="https://thanos-querier.openshift-monitoring.svc:9091"
+  export THANOS_QUERIER_URL_HOST
+  echo "route to Thanos Querier on host cluster: $THANOS_QUERIER_URL_HOST"
+  PROMETHEUS_OPERATED_URL_HOST="https://prometheus-operated-secured.$TOOLCHAIN_OPERATOR_NS_HOST.svc:8443"
+  export PROMETHEUS_OPERATED_URL_HOST
+  echo "route to Prometheus (operated) on host cluster: $PROMETHEUS_OPERATED_URL_HOST"
+  
+  # create a `grafana` serviceaccount which is allowed to access Thanos Querier and Prometheus operated (via kube-rbac-proxy)
+  oc process -f config/monitoring/grafana_serviceaccount.tmpl.yaml \
+    -p NAMESPACE="$TOOLCHAIN_OPERATOR_NS_HOST" \
+    -p NAME=grafana \
+    | oc apply -f -
+  # needed to connect to Thanos Querier
+  oc adm policy add-cluster-role-to-user cluster-monitoring-view -z grafana -n "$TOOLCHAIN_OPERATOR_NS_HOST"
+  BEARER_TOKEN_HOST="$(oc serviceaccounts get-token grafana -n $TOOLCHAIN_OPERATOR_NS_HOST)"
+  export BEARER_TOKEN_HOST
+
+   # use the 'oc create' commands along with the 'oc apply' to make sure the resources can be created or updated when they already exist
```

Comment:
correct me if I'm wrong - the `oc create` fails on conflict if you try to create an already existing resource - you should use `oc apply` instead

Reply:
hmm, the script would report an error after the first run, but it would not fail. However, `oc apply` may be a cleaner solution. Thanks for the reminder!